dbms/src/Interpreters/ClusterProxy/SelectStreamFactory.cpp

```diff
@@ -67,17 +81,34 @@ void SelectStreamFactory::createForShard(
         res.emplace_back(createLocalStream(query_ast, context, processed_stage));
     };
 
-    auto emplace_remote_stream = [&]()
+    auto emplace_remote_stream_for_database = [&]()
     {
         auto stream = std::make_shared<RemoteBlockInputStream>(shard_info.pool, query, header, context, nullptr, throttler, external_tables, processed_stage);
         stream->setPoolMode(PoolMode::GET_MANY);
         stream->setMainTable(main_table);
         res.emplace_back(std::move(stream));
     };
+    
+    auto emplace_remote_stream_for_func = [&]()
+    {
+        auto stream = std::make_shared<RemoteBlockInputStream>(shard_info.pool, query, header, context, nullptr, throttler, external_tables, processed_stage);
+        stream->setPoolMode(PoolMode::GET_MANY);
+        res.emplace_back(std::move(stream));
+    };
 
     if (shard_info.isLocal())
     {
-        StoragePtr main_table_storage = context.tryGetTable(main_table.database, main_table.table);
+        StoragePtr main_table_storage;
+        
+        if (table_func_ptr)
+        {
+            auto table_function = static_cast<ASTFunction *>(table_func_ptr.get());
+            main_table_storage = TableFunctionFactory::instance().get(table_function->name, context)->execute(table_func_ptr, context);
```

Comment:
It's better to not execute table function on initiating server.
Because it may execute successfully only on remote nodes.
For example: 
`remote('host', merge(db, '^table$'))`
- database `db` may be absent locally.